probe_src/probe_py/analysis.py

```diff
@@ -1,44 +1,46 @@
 import typing
-import networkx
+import networkx as nx
 from .parse_probe_log import ProvLog, Op, CloneOp, ExecOp, WaitOp, CLONE_THREAD
 
+# [pid, exec_epoch_no, tid, Op | str, op_index]
+Node: typing.TypeAlias = tuple[int, int, int, Op | str, int]
+program_order_edges = list[tuple[Node, Node]]()
+fork_join_edges = list[tuple[Node, Node]]()
+exec_edges = list[tuple[Node, Node]]()
 
-def construct_process_graph(process_tree_prov_log: ProvLog) -> str:
-    """
-    Construct a happens-before graph from process_tree_prov_log
-
-    The graph shows the order that prov ops happen.
-    """
-    Node: typing.TypeAlias = tuple[int, int, int, Op | str]
+def provlog_to_digraph(process_tree_prov_log: ProvLog) -> nx.digraph:
     nodes = list[Node]()
-    program_order_edges = list[tuple[Node, Node]]()
+    # program_order_edges = list[tuple[Node, Node]]()
     proc_to_ops = dict[tuple[int, int, int], list[Node]]()
-    last_exec_epoch = dict[int, int]()
-
-    for pid, process in process_tree_prov_log.processes.items():
+    last_exec_epoch = dict[int, int]()    
+    for (pid), process in process_tree_prov_log.processes.items():
         for exec_epoch_no, exec_epoch in process.exec_epochs.items():
+            # to find the last executing epoch of the process
             last_exec_epoch[pid] = max(last_exec_epoch.get(pid, 0), exec_epoch_no)
 
             # Reduce each thread to the ops we actually care about
             for tid, thread in exec_epoch.threads.items():
                 context = (pid, exec_epoch_no, tid)
-
+                ops = list[Node]()
+                
                 # Filter just the ops we are interested in
-                ops = [
-                    (*context, op.data)
-                    for op in thread.ops
-                    if isinstance(op.data, CloneOp | ExecOp | WaitOp)
-                ]
+                op_index = 0
+                
+                for op in thread.ops:
+                    if isinstance(op.data, CloneOp | ExecOp | WaitOp):
+                        ops.append((*context, op.data, op_index))
+                    op_index+=1
 
                 nodes.extend(ops)
                 program_order_edges.extend(zip(ops[:-1], ops[1:]))
-
+                
                 # Store these so we can hook up forks/joins between threads
                 proc_to_ops[context] = ops
 
     def first(pid: int, exid: int, tid: int) -> Op:
         if not proc_to_ops.get((pid, exid, tid)):
-            entry_node = (pid, exid, tid, "<entry>")
+            entry_node = (pid, exid, tid, "<entry>", -1)
```

Comment:
Remove `"<entry>"` (see comment on `Node: TypeAlias = ...`). I'll just know that if the op no was -1, then this must be a "fake" node (either entry or exit, based on its position in the list).